`@xstate/react`: Correct machine reference in the `useInterpret` hook (#2685)
statelyai/xstate: .changeset/long-icons-pull.md

---
'@xstate/react': patch
---

> 

Reviewer: ```suggestion

author: @farskid
author: @Andarist 

```

> Correct reference to the machine in `useInterpret` hook.

Reviewer: This is an implementation detail - it doesn't quite matter for our consumers so it's best to leave this out
```suggestion
```

Author: How do we word it so that the consumers are aware of a bug fix?

Reviewer: The other suggestion I've made mentions that it's a fix for a regression

> Make sure a warning is shown in the console in the development mode when machine reference is updated during the hook lifecycle. This usually happens when machine options are dependent on external values and they're passed via `withConfig`.

Reviewer: ```suggestion
Fixed a regression with a development-only warning not being shown when a machine reference is updated during the hook lifecycle. This usually happens when machine options are dependent on external values and they're passed via `withConfig`.
```

I've checked that this was introduced correctly https://github.com/statelyai/xstate/pull/918/files but regressed in https://github.com/statelyai/xstate/pull/1472/files

```js
const machine = createMachine({
  initial: 'foo',
  context: { id: 1 },
  states: {
    foo: {
      on: {
        CHECK: {
          target: 'bar',
          cond: 'hasOverflown'
        }
      }
    },
    bar: {}
  }
});

const [id, setId] = useState(1);
const [current, send] = useMachine(
  machine.withConfig({
    guards: {
      hasOverflown: () => id > 1 // id is a reference to an outside value
    }
  })
);

// later when id updates
setId(2);

> // Now the reference to `machine.withConfig` from above is updated but in the guard, the id value is still stale (id=1).

Reviewer: ```suggestion
// Now the reference passed to `useMachine` (the result of `machine.withConfig`) is updated but the interpreted machine stays the same. So the guard is still the previous one that got passed to the `useMachine` initially, and it closes over the stale `id`.
```

> ```

Reviewer: It would be great if we could also mention how one can fix this here, add a note to the docs about this and maybe even add a link to that note in the warning itself

Author: Good idea. I'm updating the docs in a separate PR. in that same PR, I'll also refer to it here. sounds good?

Reviewer: Slightly unsure about the exact shape of that other PR - but if you prefer to make those changes in a separate PR, then it's fine for me.

Author: it'll be a doc-only change. Let's take it there.
